File: ros2_ws/src/hardware_integration/src/serial_output.cpp
#include <chrono>
#include <memory>
#include <string>
#include <iostream>
#include <fstream>
#include <nlohmann/json.hpp>
#include <SerialPort.h> // Example for serial communication

#include "rclcpp/rclcpp.hpp"
#include "cascade_msgs/msg/motor_throttle.hpp"

using namespace std::chrono_literals;
using std::placeholders::_1;

class SerialOutputNode : public rclcpp::Node
{
public:
    SerialOutputNode() : Node("serial_output_node")
    {
        subscription_ = this->create_subscription<cascade_msgs::msg::MotorThrottle>(
            "/motor_throttle", 10, std::bind(&SerialOutputNode::throttle_callback, this, _1));

        // Initialize serial port (replace "COM3" with your ESP32 port and set the correct baud rate)
        serial_port_ = std::make_shared<SerialPort>("COM3", 9600);
        if (serial_port_->isOpen()) {
            RCLCPP_INFO(this->get_logger(), "Serial port opened successfully.");
        } else {
            RCLCPP_ERROR(this->get_logger(), "Failed to open serial port.");
        }
    }

private:
    void throttle_callback(const cascade_msgs::msg::MotorThrottle &msg)
    {
        // Create a JSON object following the specified format
        nlohmann::json json_msg = {
            {"actuators", {
                {"thrusters", {
                    {{"speed", msg.fli}, {"id", 1}},
                    {{"speed", msg.fri}, {"id", 2}},
                    {{"speed", msg.bli}, {"id", 3}},
                    {{"speed", msg.bri}, {"id", 4}},
                    {{"speed", msg.flo}, {"id", 5}},
                    {{"speed", msg.fro}, {"id", 6}},
                    {{"speed", msg.blo}, {"id", 7}},
                    {{"speed", msg.bro}, {"id", 8}}
                }},
                {"servos", {
                    {{"angle", 0}, {"id", 9}} // Example; adjust as needed
                }},
                {"motors", nlohmann::json::array()} // Empty array for motors
            }}
        };

        // Convert JSON object to string
        std::string json_str = json_msg.dump();

        // Send JSON string over serial
        if (serial_port_->isOpen()) {
            serial_port_->writeString(json_str + "\n");
        } else {
            RCLCPP_ERROR(this->get_logger(), "Serial port is not open.");
        }
    }

    rclcpp::Subscription<cascade_msgs::msg::MotorThrottle>::SharedPtr subscription_;
    std::shared_ptr<SerialPort> serial_port_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SerialOutputNode>());
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/hardware_integration/hardware_integration/serial_output.py
import rclpy
from rclpy.node import Node
from cascade_msgs.msg import MotorThrottle
import json
import serial  # Import pyserial

class SerialOutputNode(Node):
    def __init__(self):
        super().__init__('serial_output_node')

        # Initialize serial port (replace '/dev/ttyUSB0' with your ESP32 port and set the correct baud rate)
        try:
            self.serial_port = serial.Serial('/dev/ttyUSB0', 115200, timeout=1)
            self.get_logger().info("Serial port opened successfully.")
        except serial.SerialException as e:
            self.get_logger().error(f"Failed to open serial port: {e}")
            self.serial_port = None

        self.subscription = self.create_subscription(
            MotorThrottle,
            '/motor_throttle',
            self.throttle_callback,
            10
        )

    def throttle_callback(self, msg):
        if not self.serial_port or not self.serial_port.is_open:
            self.get_logger().error("Serial port is not open.")
            return

        # Create a JSON object following the specified format
        json_msg = {
            "actuators": {
                "thrusters": [
                    {"speed": msg.fli, "id": 1},
                    {"speed": msg.fri, "id": 2},
                    {"speed": msg.bli, "id": 3},
                    {"speed": msg.bri, "id": 4},
                    {"speed": msg.flo, "id": 5},
                    {"speed": msg.fro, "id": 6},
                    {"speed": msg.blo, "id": 7},
                    {"speed": msg.bro, "id": 8}
                ],
                "servos": [
                    {"angle": 0, "id": 9}  # Example; adjust as needed
                ],
                "motors": []
            }
        }

        self.get_logger().info(f'{json_msg}')

        # Convert JSON object to string and send it over serial
        json_str = json.dumps(json_msg)
        try:
            self.serial_port.write((json_str + '\n').encode('utf-8'))
        except serial.SerialException as e:
            self.get_logger().error(f"Error writing to serial port: {e}")

    def destroy_node(self):
        # Close serial port if open
        if self.serial_port and self.serial_port.is_open:
            self.serial_port.close()
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    node = SerialOutputNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
